File: projects/listContributingProjects.r
# Call with
#
# irule -F rules/projects/listContributingProjects.r
#
# This rule returns a JSON array of projects for which the calling user is a Manager or a Contributor
# By default, iRODS returns only the projects for which a user is authorized. If a user has read-access to a
# project, he is allowed to list the ACL's. However, this does not mean that the user is entitled to WRITE to the project.
# Therefore, we have to build some logic first to find out to which groups a user belongs, store those in a list and
# filter the resultset on modify rights for the groups in that list (= COLL_ACCESS_USER_ID in (*groups) ).
# Example use-case:
# - user X is member of group Y.
# - project 0001 has readonly access for group Y
# - the user should not be able to select project 0001 to ingest data
#
listContributingProjects {
    *json_str = '[]';
    *size = 0;
    *groups = '';

    userNameToUserId($userNameClient, *userId);

    # Create a list of group-IDs (the user-ID is also a "group-ID")
    foreach ( *Row2 in SELECT USER_GROUP_ID where USER_ID = *userId ) {
         *groupID = "'" ++ *Row2.USER_GROUP_ID ++ "'";
         *groups= *groups ++ "," ++ *groupID;
    }

    # Remove first comma
    *groups=substr(*groups, 1, strlen(*groups));

    # Create GenQuery since ordinary select statement cannot deal with "in (*groups)" construction
    msiMakeGenQuery("COLL_NAME", "COLL_ACCESS_NAME in ('own', 'modify object') and COLL_ACCESS_USER_ID in (*groups) and COLL_PARENT_NAME = '/nlmumc/projects'", *Query);
    msiExecGenQuery(*Query, *QOut);

    # Loop over SQL result and generate JSON array with project info
    foreach ( *Row in *QOut) {
        uuChopPath(*Row.COLL_NAME, *collection, *project);

        # Get project details
        getCollectionAVU("/nlmumc/projects/*project","title",*title,"no-title-AVU-set","false");
        listProjectContributors(*project, 'false', *contributors);
        listProjectManagers(*project, *managers);
        listProjectViewers(*project, 'false', *viewers);

        msiString2KeyValPair("", *kvp);
        msiAddKeyVal(*kvp, 'project', *project);
        msiAddKeyVal(*kvp, 'title', *title);
        msiAddKeyVal(*kvp, 'contributors', *contributors);
        msiAddKeyVal(*kvp, 'managers', *managers);
        msiAddKeyVal(*kvp, 'viewers', *viewers);

        *o = ""
        msi_json_objops(*o, *kvp, "set");

        msi_json_arrayops(*json_str, *o, "add", *size);
    }
    writeLine("stdout", *json_str);
}

INPUT *token=""
OUTPUT ruleExecOut

File: python/list_contributing_projects.py
@make(inputs=[0], outputs=[1], handler=Output.STORE)
def list_contributing_project(ctx, show_service_accounts):
    """
    Query the list of ACL for a project for the client user

    Parameters
    ----------
    ctx : Context
        Combined type of a callback and rei struct.
    show_service_accounts: str
        'true'/'false' expected; If true, hide the service accounts in the result

    Returns
    -------
    dict
        The list of usernames for managers, contributors and viewers.
        Returns an empty list if the user is not a contributor.
    """
    projects = []
    groups = ""
    username = ctx.callback.get_client_username('')["arguments"][0]

    ret = ctx.callback.userNameToUserId(username, "*userId")
    user_id = ret["arguments"][1]

    for result in row_iterator("USER_GROUP_ID",
                                  "USER_ID = '{}'".format(user_id),
                                  AS_LIST,
                                  ctx.callback):
        group_id = "'" + result[0] + "'"
        groups = groups + "," + group_id

    # Remove first comma
    groups = groups[1:]

    # Get the collection size on each resources
    parameters = "COLL_NAME"
    conditions = "COLL_ACCESS_NAME in ('own', 'modify object') " \
                 "and COLL_ACCESS_USER_ID in ({}) " \
                 "and COLL_PARENT_NAME = '/nlmumc/projects'".format(groups)

    for collection_result in row_iterator(parameters, conditions, AS_LIST, ctx.callback):
        project = {"id": collection_result[0].split("/")[3]}

        # List Contributors
        ret = ctx.callback.listProjectContributors(project["id"], "false", "")["arguments"][2]
        project["contributors"] = json.loads(ret)

        # List Managers
        ret = ctx.callback.list_project_managers(project["id"], show_service_accounts, "")["arguments"][2]
        project["managers"] = json.loads(ret)

        # List Viewers
        ret = ctx.callback.listProjectViewers(project["id"], "false", "")["arguments"][2]
        project["viewers"] = json.loads(ret)

        # Get project metadata
        # Note: Retrieving the rule outcome is done with '["arguments"][2]'
        project["title"] = ctx.callback.getCollectionAVU(collection_result[0], "title", "", "", "true")["arguments"][2]
        project["resource"] = ctx.callback.getCollectionAVU(collection_result[0], "resource", "", "", "true")["arguments"][2]

        projects.append(project)

    return projects
